barkeep_server.rb

```diff
@@ -230,6 +237,89 @@ def ensure_required_params(*required_params)
+    provider = LDAP_PROVIDERS[params[:provider_id].to_i]
+    halt 400, "LDAP provider not found." unless provider
+    erb :ldap_signin, :locals => { :provider_id => params[:provider_id], :name => provider[:name] || provider[:host] }
+  end
+
+  post "/signin/ldap_authenticate" do
+    provider = LDAP_PROVIDERS[params[:provider_id].to_i]
+    halt 400, "LDAP provider not found." unless provider
+
+    begin
+      Timeout::timeout(5) do
+        conn_params = { :host => provider[:host], :port => provider[:port] }
+        auth_params = { :method => provider[:method] }
+        
+        if provider.has_key?(:username)
+          # Here we use separate credentials, i.e. those in the config file, to query
+          # the LDAP server in order to locate the correct user entry.
+          conn = Net::LDAP.new(conn_params.merge({
+                                 :auth => auth_params.merge({
+                                            :username => provider[:username],
+                                            :password => provider[:password] })}))
+
+          if conn.bind
+            # Using the field name provided in the config file, search for the entry
+            # whose value in that field is equal to the username provided on the web
+            # form.
+            filter = Net::LDAP::Filter.eq(provider[:uid], params[:username])
+            entries = conn.search(:base => provider[:base], :filter => filter)
+            if entries.nil? or entries.empty?
+              halt 401, "Invalid username or password"
+            elsif entries.size > 1
+              # TODO log error message for duplicate user state
+              halt 401, "LDAP misconfigured"
+            end
+
+            conn2 = Net::LDAP.new(conn_params.merge({
+                                   :auth => auth_params.merge({
+                                              :username => entries.first.dn,
+                                              :password => params[:password] })}))
+
+            if conn2.bind
+              start_session(entries.first.cn.first, entries.first.mail.first)
+              session[:login_started_url]
+            else
+              halt 401, "Invalid username or password"
+            end
+          else
+            halt 401, "Barkeep could not bind as query user"
+          end
+        else
+          # This is the simpler form of LDAP authentication, wherein we simply
+          # attempt to bind using the credentials provided on the web form in order
+          # to validate the user.
+          user_dn = "#{provider[:uid]}=#{params[:username]},#{provider[:base]}\n"
+          conn = Net::LDAP.new(conn_params.merge({
+                                 :auth => auth_params.merge({
+                                            :username => user_dn,
+                                            :password => params[:password] })}))
+
+          if conn.bind
+            filter = Net::LDAP::Filter.contains(provider[:uid], params[:username])
```

Comment:
This Filter.contains needs to actually be Filter.eq.   If, for instance, you happen to have Alice Smith (asmith) and Adam Smithson (asmithson) then Alice won't be able to log in because this seach will return both, and the entries.size > 1 error condition a few line below will activate.